### PR TITLE
CO-1475: add validation of hidden field when default value is missing

### PR DESCRIPTION
**Pull request by carsonreinke**
https://todos.internet2.edu/browse/CO-1475

### CODE REVIEW

File: app/Model/CoEnrollmentAttribute.php [hunk 1]
```diff
@@ -112,8 +112,42 @@ class CoEnrollmentAttribute extends AppModel {
...
+   * Validate that the field is allowed to be hidden based on the attribute and default value
+   *
+   * @param  array Array of fields to validate
+   * @return boolean True if allowed, false if not
+   */
+  public function validateHidden($a) {
+    // Hidden attribute is only applicable to CO Person Role attributes (type 'r'), CO Person attributes (type 'g'),
+    // Organizational Identity attributes (type 'o'), or Extended Attributes (type 'x')
+    $attrCode = explode(':', $this->data['CoEnrollmentAttribute']['attribute']);
+    if($attrCode[0] != 'r' && $attrCode[0] != 'g' && $attrCode[0] != 'o' && $attrCode[0] != 'x') {
```
Comment (Jeanesily) [inline app/Model/CoEnrollmentAttribute.php:139]: Nit picky - but would in_array work here instead of stringing several checks together?

Comment (carsonreinke) [reply app/Model/CoEnrollmentAttribute.php]: @Jeanesily yep, I saw this somewhere else, but the `in_array` thing is way more readable.

**Comment by boshrin** [comment]
Some comments:

1. `ruleName` and `ruleName2` aren't all that descriptive.
1. `required` validation rules should be accompanied by `allowEmpty` rules (with an inverse value) due to technicalities in the framework. I'm not sure what it means that two rules have opposite required values... (`required` means "this key must be in the form data", `allowEmpty` means "the value for this key may be empty".)
1. Unlocalized strings should not be hardcoded in validation rules. (There are some legacy instances that haven't been clean up yet, but those shouldn't serve as examples.) Instead, the validation rule should return a localized text string on failure. See `AppModel::validateInput()` for an example.

**Comment by boshrin** [comment]
Also, I'm not sure why this was flagged for v4.0.0 rather than v3.3.2, so this PR should probably be rebased against hotfix-3.3.x.

**Comment by Jeanesily** [comment]
@boshrin Closing this out as I recreated/rebased it here: https://github.com/Internet2/comanage-registry/pull/182 (including the original commit: https://github.com/Internet2/comanage-registry/pull/182/commits/13adaba258dee8c74ad1141c391da751a7f94d20)

**Comment by Jeanesily** [comment]
Oops, actually I can't close this @boshrin  if you could do that, thanks!

**Comment by boshrin** [comment]
Closing per @Jeanesily

**Pull request closed without merging**